lib/Jsrt/JsrtDebugManager.cpp

```diff
@@ -0,0 +1,693 @@
+        tempScriptContext = tempScriptContext->next)
+    {
+        tempScriptContext->GetDebugContext()->GetProbeContainer()->AsyncDeactivate();
+    }
+
+    if (Js::Configuration::Global.EnableJitInDebugMode())
+    {
+        scriptContext->GetThreadContext()->GetDebugManager()->GetDebuggingFlags()->SetForceInterpreter(false);
+    }
+}
+
+Js::DynamicObject* JsrtDebugManager::GetScript(Js::Utf8SourceInfo* utf8SourceInfo)
+{
+    Js::DynamicObject* scriptObject = utf8SourceInfo->GetScriptContext()->GetLibrary()->CreateObject();
+
+    JsrtDebugUtils::AddScriptIdToObject(scriptObject, utf8SourceInfo);
+    JsrtDebugUtils::AddFileNameOrScriptTypeToObject(scriptObject, utf8SourceInfo);
+    JsrtDebugUtils::AddLineCountToObject(scriptObject, utf8SourceInfo);
+    JsrtDebugUtils::AddPropertyToObject(scriptObject, JsrtDebugPropertyId::sourceLength, utf8SourceInfo->GetCchLength(), utf8SourceInfo->GetScriptContext());
+
+    return scriptObject;
+}
+
+Js::JavascriptArray* JsrtDebugManager::GetScripts(Js::ScriptContext* scriptContext)
+{
+    Js::JavascriptArray* scriptsArray = scriptContext->GetLibrary()->CreateArray();
+
+    int index = 0;
+
+    for (Js::ScriptContext *tempScriptContext = scriptContext->GetThreadContext()->GetScriptContextList();
+    tempScriptContext != nullptr && !tempScriptContext->IsClosed();
+        tempScriptContext = tempScriptContext->next)
+    {
+        tempScriptContext->GetSourceList()->Map([&](int i, RecyclerWeakReference<Js::Utf8SourceInfo>* utf8SourceInfoWeakRef)
+        {
+            Js::Utf8SourceInfo* utf8SourceInfo = utf8SourceInfoWeakRef->Get();
+            if (utf8SourceInfo != nullptr && !utf8SourceInfo->GetIsLibraryCode() && utf8SourceInfo->HasDebugDocument())
+            {
+                bool isCallerLibraryCode = false;
+
+                bool isDynamic = utf8SourceInfo->IsDynamic();
+
+                if (isDynamic)
+                {
+                    // If the code is dynamic (eval or new Function) only return the script if parent is non-library
+                    Js::Utf8SourceInfo* callerUtf8SourceInfo = utf8SourceInfo->GetCallerUtf8SourceInfo();
+
+                    while (callerUtf8SourceInfo != nullptr && !isCallerLibraryCode)
+                    {
+                        isCallerLibraryCode = callerUtf8SourceInfo->GetIsLibraryCode();
+                        callerUtf8SourceInfo = callerUtf8SourceInfo->GetCallerUtf8SourceInfo();
+                    }
+                }
+
+                if (!isCallerLibraryCode)
+                {
+                    Js::DynamicObject* sourceObj = this->GetScript(utf8SourceInfo);
+
+                    if (sourceObj != nullptr)
+                    {
+                        Js::Var marshaledObj = Js::CrossSite::MarshalVar(scriptContext, sourceObj);
+                        Js::JavascriptOperators::OP_SetElementI((Js::Var)scriptsArray, Js::JavascriptNumber::ToVar(index, scriptContext), marshaledObj, scriptContext);
```

Comment:
nice catch, fixed

---
In reply to: [62755905](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62755905) [](ancestors = 62755905)